libraries/lib-ipc/IPCServer.cpp

```diff
@@ -0,0 +1,116 @@
+#include "IPCServer.h"
+#include "IPCChannel.h"
+
+#include <atomic>
+#include <thread>
+
+#include "internal/ipc-types.h"
+#include "internal/socket_guard.h"
+#include "internal/BufferedIPCChannel.h"
+
+class IPCServer::Impl
+{
+   std::atomic<bool> mAlive{true};
+   std::unique_ptr<std::thread> mWorker;
+public:
+
+   ~Impl()
+   {
+      mAlive = false;
+      if(mWorker)
+         mWorker->join();
+   }
+
+   bool CreateChannel(int id, IPCChannelStatusCallback* callback)
+   {
+      if(mWorker)
+         return false;
+
+      auto serverfd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+      if(serverfd == INVALID_SOCKET)
+         return false;
+
+      socket_guard serverfd_guard { serverfd };
+
+      sockaddr_in addrin {};
+      addrin.sin_family = AF_INET;
+      addrin.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+      addrin.sin_port = htons(static_cast<u_short>(id));
+
+      if(bind(serverfd, reinterpret_cast<const sockaddr*>(&addrin), sizeof(addrin)) == SOCKET_ERROR)
+         return false;
+
+      if(listen(serverfd, 1) == SOCKET_ERROR)
+         return false;
+
+      mWorker = std::make_unique<std::thread>([this, id, callback, serverfd, guard = std::move(serverfd_guard)]()
+      {
+#ifdef _WIN32
+         WSAPOLLFD pollfd{};
+#else
+         struct pollfd pollfd{};
+#endif
+
+         pollfd.events = POLLRDNORM;
+         pollfd.fd = serverfd;
+
+         SOCKET clientfd {INVALID_SOCKET};
+
+         while(true)
+         {
+#ifdef _WIN32
+            auto res = WSAPoll(&pollfd, 1, 0);
+#else
+            auto res = poll(&pollfd, 1, 0);
+#endif
+            if(!mAlive)
+               //do not notify
+               return;
+            if(res == 0)
+            {
+               std::this_thread::yield();
+               continue;
+            }
+            if (res == 1)
+               clientfd = accept(serverfd, nullptr, nullptr);
```

Comment:
`man poll` tells me that -1 is a possible return value, indicating an error, stored in `errno` (which is a macro, not really a variable, on my system:  I suppose they addressed the thread safety problem).

So please make the check more exhaustive.